utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,184 @@
+///
+/// assert_eq!(zv_no.get(0).as_deref(), Some(&None));
+/// assert_eq!(zv_no.get(1).as_deref(), Some(&NonZeroU8::new(1)));
+/// assert_eq!(zv_no.get(2).as_deref(), Some(&NonZeroU8::new(2)));
+/// assert_eq!(zv_no.get(3).as_deref(), Some(&None));
+/// ```
+// Invariants:
+// The union stores [`NicheBytes::INVALID_BIT_PATTERN`] when None.
+// Any other bit pattern is a valid.
+#[repr(packed)]
+pub union NichedOptionULE<U: NicheBytes<N> + ULE, const N: usize> {
+    invalid: [u8; N],
+    valid: U,
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> NichedOptionULE<U, N> {
+    /// New NichedOptionULE<U, N> from Option<U>
+    pub fn new(opt: Option<U>) -> Self {
+        assert!(N == core::mem::size_of::<U>());
+        match opt {
+            Some(u) => Self { valid: u },
+            None => Self {
+                invalid: <U as NicheBytes<N>>::INVALID_BIT_PATTERN,
+            },
+        }
+    }
+
+    /// Convert to an Option<U>
+    pub fn get(self) -> Option<U> {
+        if unsafe { self.invalid } == <U as NicheBytes<N>>::INVALID_BIT_PATTERN {
+            None
+        } else {
+            // Safety: the inner value is valid
+            unsafe { Some(self.valid) }
+        }
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> From<Option<U>> for NichedOptionULE<U, N> {
+    fn from(opt: Option<U>) -> Self {
+        Self::new(opt)
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> From<NichedOptionULE<U, N>> for Option<U> {
+    fn from(noule: NichedOptionULE<U, N>) -> Self {
+        NichedOptionULE::get(noule)
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> Copy for NichedOptionULE<U, N> {}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> Clone for NichedOptionULE<U, N> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<U: NicheBytes<N> + ULE + PartialEq, const N: usize> PartialEq for NichedOptionULE<U, N> {
+    fn eq(&self, other: &Self) -> bool {
+        self.get().eq(&other.get())
+    }
+}
+
+impl<U: NicheBytes<N> + ULE + Eq, const N: usize> Eq for NichedOptionULE<U, N> {}
+
+/// Safety for ULE trait
+/// 1. NichedOptionULE does not have any padding bytes due to `#[repr(packed)]` on a struct
+///    containing only ULE fields.
+///    NichedOptionULE either contains INVALID_BIT_PATTERN or valid U byte sequences.
+///    In both cases the data is initialized.
+/// 2. NichedOptionULE is aligned to 1 byte due to `#[repr(packed)]` on a struct containing only
+///    ULE fields.
+/// 3. validate_byte_slice impl returns an error if invalid bytes are encountered.
+/// 4. validate_byte_slice impl returns an error there are extra bytes.
+/// 5. The other ULE methods are left to their default impl.
+/// 6. NichedOptionULE equality is based on ULE equality of the subfield.
+unsafe impl<U: NicheBytes<N> + ULE, const N: usize> ULE for NichedOptionULE<U, N> {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), crate::ZeroVecError> {
+        let size = size_of::<Self>();
```

Comment:
I think we need to assert that N == sizeof(U) here as well

Reply:
though I guess if the size is incorrect it will always fail one or the other chunks bounds. Perhaps we can just mention that in a comment and include a debug assertion?

Reply:
Added a debug assert.